File: python_sender/magxxic/core/proxy_validator.py
import socks
import socket
from urllib.parse import urlparse
from concurrent.futures import ThreadPoolExecutor

def check_proxy(proxy_str, test_host="8.8.8.8", test_port=53, timeout=5):
    """
    Check if a SOCKS5 proxy is working by attempting a connection to a reliable host.
    """
    if not proxy_str:
        return False

    if not proxy_str.startswith("socks5://"):
        proxy_url = f"socks5://{proxy_str}"
    else:
        proxy_url = proxy_str

    try:
        parsed = urlparse(proxy_url)
        s = socks.socksocket()
        s.set_proxy(
            proxy_type=socks.SOCKS5,
            addr=parsed.hostname,
            port=parsed.port,
            username=parsed.username,
            password=parsed.password
        )
        s.settimeout(timeout)
        s.connect((test_host, test_port))
        s.close()
        return True
    except Exception:
        return False

def validate_proxies(proxies, max_workers=20, test_smtp=False):
    """
    Validate a list of proxies concurrently and return only the working ones.
    If test_smtp is True, it tests port 25 connectivity.
    """
    if not proxies:
        return []

    # Use port 25 for testing if requested, otherwise 53 (DNS)
    test_port = 25 if test_smtp else 53
    test_host = "smtp.google.com" if test_smtp else "8.8.8.8"

    working_proxies = []
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        future_to_proxy = {executor.submit(check_proxy, p, test_host=test_host, test_port=test_port): p for p in proxies}
        for future in future_to_proxy:
            proxy = future_to_proxy[future]
            try:
                if future.result():
                    working_proxies.append(proxy)
            except Exception:
                pass
    return working_proxies

File: magxxic/core_js/proxy_validator.js
const { SocksProxyAgent } = require('socks-proxy-agent');
const net = require('net');

async function checkProxy(proxyStr, testHost = "8.8.8.8", testPort = 53, timeout = 5000) {
    if (!proxyStr) return false;

    let proxyUrl = proxyStr;
    if (!proxyStr.startsWith("socks5://")) {
        proxyUrl = `socks5://${proxyStr}`;
    }

    const agent = new SocksProxyAgent(proxyUrl);

    return new Promise((resolve) => {
        const socket = agent.callback(
            { host: testHost, port: testPort },
            { timeout }
        );

        socket.on('connect', () => {
            socket.destroy();
            resolve(true);
        });

        socket.on('error', () => {
            resolve(false);
        });

        setTimeout(() => {
            socket.destroy();
            resolve(false);
        }, timeout);
    });
}

async function validateProxies(proxies, maxWorkers = 20) {
    if (!proxies || proxies.length === 0) return [];

    const workingProxies = [];
    const chunks = [];
    for (let i = 0; i < proxies.length; i += maxWorkers) {
        chunks.push(proxies.slice(i, i + maxWorkers));
    }

    for (const chunk of chunks) {
        const results = await Promise.all(chunk.map(p => checkProxy(p)));
        results.forEach((res, i) => {
            if (res) workingProxies.push(chunk[i]);
        });
    }
    return workingProxies;
}

module.exports = { checkProxy, validateProxies };
